app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsSalableWithReservationsCondition.php

```diff
@@ -84,21 +102,31 @@ public function execute(string $sku, int $stockId, float $requestedQty): Product
             return $this->productSalableResultFactory->create(['errors' => $errors]);
         }
 
-        $qtyWithReservation = $stockItemData[GetStockItemDataInterface::QUANTITY] +
-            $this->getReservationsQuantity->execute($sku, $stockId);
-        /** @var StockItemConfigurationInterface $stockItemConfiguration */
-        $stockItemConfiguration = $this->getStockItemConfiguration->execute($sku, $stockId);
-        $qtyLeftInStock = $qtyWithReservation - $stockItemConfiguration->getMinQty() - $requestedQty;
-        $isEnoughQty = (bool)$stockItemData[GetStockItemDataInterface::IS_SALABLE] && $qtyLeftInStock >= 0;
-        if (!$isEnoughQty) {
+        $productType = $this->getProductTypesBySkus->execute([$sku])[$sku];
+
+        if (true === $this->isSourceItemsAllowedForProductType->execute($productType)) {
+            $qtyWithReservation = $stockItemData[GetStockItemDataInterface::QUANTITY] +
+                $this->getReservationsQuantity->execute($sku, $stockId);
+
+            /** @var StockItemConfigurationInterface $stockItemConfiguration */
+            $stockItemConfiguration = $this->getStockItemConfiguration->execute($sku, $stockId);
+            $qtyLeftInStock = $qtyWithReservation - $stockItemConfiguration->getMinQty() - $requestedQty;
+            $isSalable = (bool)$stockItemData[GetStockItemDataInterface::IS_SALABLE];
+            $isSalableForRequestedQty = $isSalable && $qtyLeftInStock >= 0;
+        } else {
+            $isSalableForRequestedQty = (bool)$stockItemData[GetStockItemDataInterface::IS_SALABLE];
```

Comment:
Need to replace on Exception
Because it is unexpected and  ambiguous behavior to check quantity for complex products